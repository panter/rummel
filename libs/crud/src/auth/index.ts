export * from './crud-authorization';
export * from './crud-resource.decorator';
