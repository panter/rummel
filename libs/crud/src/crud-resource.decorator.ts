import { SetMetadata } from '@nestjs/common';

export const CRUD_RESOURCE = 'CrudResource';

export const CrudResource = (resourceName: string) =>
  SetMetadata(CRUD_RESOURCE, resourceName);
