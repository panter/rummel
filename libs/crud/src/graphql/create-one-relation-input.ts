import { Type } from '@nestjs/common';
import { ArgsType, Field, InputType } from '@nestjs/graphql';
import { ConnectRelationInput } from './generic-types';
import { typesCache } from './types-cache';
import { upsertInput } from './upsert-input';
import { getTypeName } from './utils';

export const createOneRelationInput = <T>(
  classRef: Type<T>,
  options?: {
    parentRef: Type<any>;
    hideCreate?: boolean;
    parentProperty?: string;
  },
) => {
  const typeName = getTypeName(classRef);
  const withoutTypeName = options?.parentRef
    ? `Without${getTypeName(options?.parentRef)}`
    : '';
  const name = `${typeName}CreateNestedOne${withoutTypeName}Input`;

  if (typesCache[name]) {
    return typesCache[name];
  }

  @ArgsType()
  @InputType(name, { isAbstract: false })
  class RelationInputArgsType {
    @Field(() => ConnectRelationInput, { nullable: true })
    connect: any;

    create: any;
  }

  typesCache[name] = RelationInputArgsType;

  if (!options?.hideCreate) {
    // call createInput after adding to cache because of recursive call
    const WhereInputType = upsertInput(classRef, {
      ignoreType: options?.parentRef,
    });
    Field(() => WhereInputType, { nullable: true })(
      RelationInputArgsType.prototype,
      'create',
    );
  }

  return RelationInputArgsType;
};
