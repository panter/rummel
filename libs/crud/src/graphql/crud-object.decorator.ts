import { EntityManager } from '@mikro-orm/postgresql';
import { Type } from '@nestjs/common';
import 'reflect-metadata';
import { getTypeName } from './utils';

export type InputResolverOptions<Entity> = {
  em: EntityManager;
  currentUser: any;
  currentOrmData?: Entity;
  type: Type<Entity>;
};
export type InputResolver<Entity = any, GqlInput = any> = (
  gqlInput: GqlInput,
  options: InputResolverOptions<Entity>,
) => Promise<void>;

/**
 * The idea is to cache the options for each type.
 * Ideally we would like to use Reflect.getMetadata but it
 * I ran into problems so i resorted to to a classic map cache approach.
 *
 */

const cache = new Map<Type<unknown>, CrudObjectOptions>();
const addToCache = (target: any, options: CrudObjectOptions) => {
  cache.set(target, options);
};

// const CRUD_OBJECT_METADATA_NAME = 'crud:object';

export type CrudObjectProps<Entity = any, GqlInput = any> = {
  alias?: Type<unknown>;
  createOne?: [InputResolver<Entity, GqlInput>];
  updateOne?: [InputResolver<Entity, GqlInput>];
  // findManyOne?: (all the necessary args) => void );
  // ... all the cruds overwride
};
export type CrudObjectOptions<Entity = any, GqlInput = any> = CrudObjectProps<
  Entity,
  GqlInput
> & {
  targetName: string;
  target: Type<unknown>;
};

export function CrudObject<Entity = any, GqlInput = any>(
  options: CrudObjectProps<Entity, GqlInput> = {},
) {
  return function (target: any) {
    const meta: CrudObjectOptions<Entity, GqlInput> = {
      ...options,
      target,
      targetName: getTypeName(target),
    };
    addToCache(target, meta);
    // Reflect.defineMetadata(
    //   CRUD_OBJECT_METADATA_NAME,
    //   meta,
    // );
  };
}

export function getCrudObjectOptions<DTO>(
  target: Type<DTO>,
): CrudObjectOptions | undefined {
  return cache.get(target);
  // const options = getPrototypeChain(target).reduce((options, Cls) => {
  //   if (cache.get(Cls.prototype)) {
  //     return cache.get(Cls.prototype);
  //   }
  //   return options;
  // }, [] as any);

  // return options;
}
