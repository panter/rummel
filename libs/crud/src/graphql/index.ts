export * from './upsert-input';
export * from './resolvers/create-one.resolver';
export * from './resolvers/update-one.resolver';
export * from './resolvers/find-one.resolver';
export * from './resolvers/find-many.resolver';
export * from './resolvers/delete-one.resolver';
export * from './resolvers/object-relation-resolver';
export * from './create-one-entity-args';
export * from './find-many-entity-args';
export * from './gql-filter-to-mikro-orm';
export * from './gql-upsert-input-to-mikro-orm';
export * from './generic-types';
export * from './crud-field.decorator';
