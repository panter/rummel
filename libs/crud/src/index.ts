export * from './graphql';
export * from './crud-resource.decorator';
export * from './crud.service';
export * from './crud.module';
