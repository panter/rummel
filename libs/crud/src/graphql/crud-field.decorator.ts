import { EntityManager, ObjectQuery } from '@mikro-orm/postgresql';
import { Type } from '@nestjs/common';
import { ReturnTypeFunc } from '@nestjs/graphql';
import { TypeOptions } from '@nestjs/graphql/dist/interfaces/type-options.interface';
import { uniqBy } from 'lodash';
import 'reflect-metadata';
// import { UserEntity } from '../../modules/user/models/user.entity';
import { CrudInfo, getPrototypeChain } from './utils';
import { AuthenticatedUser } from './types';

/**
 * The idea is to cache the options for each type.
 * Ideally we would like to use Reflect.getMetadata but it
 * I ran into problems so i resorted to to a classic map cache approach.
 *
 */

const cache = new Map<Type<unknown>, { [key: string]: CrudFieldOptions }>();
const addToCache = (target: any, key: string, options: any) => {
  const cached = cache.get(target);
  if (cached) {
    cached[key] = options;
  } else {
    cache.set(target, { [key]: options });
  }
};

// const CRUD_FIELD_METADATA_NAME = 'crud:field';
export type WhereFieldResolver<Entity = any, GqlWhere = any> = (
  ormQuery: ObjectQuery<Entity>,
  gqlWhere: GqlWhere,
  options: {
    currentUser?: AuthenticatedUser | null;
    key: string | number | symbol;
  },
) => void;

export type InputFieldResolverOptions<Entity> = {
  em: EntityManager;
  currentUser?: AuthenticatedUser | null;
  currentOrmData?: Entity;
  rootOrmData?: Entity;
  type: Type<Entity>;
  crudInfo: CrudInfo;
  skipInputResolver?: boolean;
  key: string | number | symbol;
};
export type InputFieldResolver<Entity = any, GqlInput = any> = (
  ormData: Entity,
  gqlInput: GqlInput,
  options: InputFieldResolverOptions<Entity>,
) => Promise<void>;

export type CrudRelationFieldOptions = {
  /**
   * If set to `true`, the relation cannot be connected by primary key
   */
  hideConnect?: boolean;
  /**
   *  If set to `true`, the relation can be created with all it's fields.
   */
  showCreate?: boolean;
  /**
   *  If set to `true`, the relation can be updated with all it's fields.
   */
  showUpdate?: boolean;
  /**
   *  If set to `true`, the relation cannot be disconnected by primary key.
   */
  hideDisconnect?: boolean;
};

export type CrudFieldProps<Entity = any, GqlInput = any, GqlWhere = any> = {
  typeFn?: ReturnTypeFunc;
  // TODO: this is for CatalogProductPropertyWithValue so we can handle it like the other fields
  ormTypeFn?: ReturnTypeFunc;
  name?: string;
  /**
   * Schema instructions.
   *  If set to `true`, the field will be hidden from the create operation, so the field cannot be set during creation.
   */
  hideCreate?: boolean;
  /**
   * If set to `true`, the field will be hidden from the update operation, so the field cannot be updated.
   */
  hideUpdate?: boolean;
  /**
   * Schema instructions.
   * If set to `true`, the field will be hidden from the where input type, so the field cannot be filtered.
   */
  hideWhere?: boolean;
  /**
   * Schema instructions.
   * If set to `true`, the relation can be connected by primary key
   */
  fieldOptions?: TypeOptions;
  /**
   * Relation schema instructions
   */
  relation?: CrudRelationFieldOptions & {
    parentPropert?: string;
  };
  staticWhereResolver?: WhereFieldResolver<Entity, GqlWhere>;
  whereResolver?: WhereFieldResolver<Entity, GqlWhere>;
  inputResolver?: true | InputFieldResolver<Entity, GqlInput>;
  staticInputResolver?: InputFieldResolver<Entity, GqlInput>;
};
export type CrudFieldOptions<
  Entity = any,
  GqlInput = any,
  GqlWhere = any,
> = CrudFieldProps<Entity, GqlInput, GqlWhere> & { propertyKey: string };

export function CrudField<Entity = any, GqlInput = any, GqlWhere = any>(
  options: CrudFieldProps<Entity, GqlInput, GqlWhere> = {},
) {
  return function (target: any, propertyKey: string) {
    const meta: CrudFieldOptions<Entity, GqlInput, GqlWhere> = {
      name: propertyKey,
      ...options,
      propertyKey,
    };

    addToCache(target, propertyKey, meta);
    // Reflect.defineMetadata(
    //   CRUD_FIELD_METADATA_NAME,
    //   { name: propertyKey, ...options, propertyKey },
    //   target,
    //   propertyKey
    // );
  };
}

export function getCrudFields<DTO>(target: Type<DTO>): CrudFieldOptions[] {
  const options = getPrototypeChain(target).reduce((fields, Cls) => {
    const cached = cache.get(Cls.prototype);

    if (!cached) {
      return fields;
    }
    return uniqBy(
      Object.values(cached).reduce((allFields, field) => {
        return [...allFields, field];
      }, fields),
      'name',
    );
  }, [] as any);

  return options;
}
