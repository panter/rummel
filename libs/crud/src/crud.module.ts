import { ConfigModule } from '@nestjs/config';
import { DiscoveryModule } from '@nestjs/core';
import { DynamicModule, Module, OnModuleInit } from '@nestjs/common';
import { CrudAuditCallback, CrudAuthorizeCallback } from './';
import { CrudService } from './service/crud.service';
import { ModuleAsyncOptions } from '@panter/nestjs-utils';
import { CrudEntityServiceFactory } from './service/crud-entity-service.factory';
import { CrudEntityService } from '.';

export interface CrudModuleOptions<T> {
  defaultRelationModifier?: boolean;
  authorizeCallback?: CrudAuthorizeCallback<T>;
  auditCallback?: CrudAuditCallback<T>;
}

export type CrudModuleAsyncOptions<T> = ModuleAsyncOptions<
  CrudModuleOptions<T>
> & { defaultRelationModifier?: boolean };

let relationModifier: boolean | undefined = true;

export const defaultRelationModifier = () => relationModifier;

@Module({})
export class CrudModule implements OnModuleInit {
  static async forRootAsync<T>({
    defaultRelationModifier,
    inject,
    imports,
    useFactory,
  }: CrudModuleAsyncOptions<T>): Promise<DynamicModule> {
    relationModifier = defaultRelationModifier;
    return {
      module: CrudModule,
      imports: [...(imports || []), ConfigModule, DiscoveryModule],
      providers: [
        {
          provide: 'CONFIG_OPTIONS',
          useFactory: async (...args: any[]) => await useFactory(...args),
          inject: inject || [],
        },
        CrudService,
        CrudEntityService,
        CrudEntityServiceFactory,
      ],
      exports: [
        'CONFIG_OPTIONS',
        CrudService,
        CrudEntityService,
        CrudEntityServiceFactory,
      ],
    };
  }

  constructor(private readonly curdAuthorizationService: CrudService) {}

  onModuleInit() {
    this.curdAuthorizationService.init();
  }
}
