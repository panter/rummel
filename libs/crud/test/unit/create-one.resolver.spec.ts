import {
  GraphQLSchemaBuilderModule,
  GraphQLSchemaFactory,
  Resolver,
} from '@nestjs/graphql';
import { Test, TestingModule } from '@nestjs/testing';
import { printSchema } from 'graphql/utilities';
import { CreateOneResolver } from '../../src';
import { Company } from '../fixtures/company.entity';
import { Group } from '../fixtures/group.entity';
import { User } from '../fixtures/user.entity';

@Resolver(() => User)
export class CreateOneUserResolver extends CreateOneResolver(User) {}

@Resolver(() => Company)
export class CreateOneCompanyResolver extends CreateOneResolver(Company) {}

@Resolver(() => Group)
export class CreateOneGroupResolver extends CreateOneResolver(Group) {}

describe('CreateOneResolver', () => {
  let app: TestingModule;

  beforeEach(async () => {
    app = await Test.createTestingModule({
      imports: [GraphQLSchemaBuilderModule],
    }).compile();
  });

  it('should generate graphql schema for CreateOneUserResolver', async () => {
    const gqlSchemaFactory = app.get(GraphQLSchemaFactory);
    const schema = await gqlSchemaFactory.create([CreateOneUserResolver], {
      skipCheck: true, //to avoid QueryRoot type must be provided error
    });
    expect(printSchema(schema)).toMatchSnapshot();
  });

  it('should generate graphql schema for CreateOneCompanyResolver', async () => {
    const gqlSchemaFactory = app.get(GraphQLSchemaFactory);
    const schema = await gqlSchemaFactory.create([CreateOneCompanyResolver], {
      skipCheck: true, //to avoid QueryRoot type must be provided error
    });
    expect(printSchema(schema)).toMatchSnapshot();
  });

  it('should generate graphql schema for CreateOneGroupResolver', async () => {
    const gqlSchemaFactory = app.get(GraphQLSchemaFactory);
    const schema = await gqlSchemaFactory.create([CreateOneGroupResolver], {
      skipCheck: true, //to avoid QueryRoot type must be provided error
    });
    expect(printSchema(schema)).toMatchSnapshot();
  });
});
