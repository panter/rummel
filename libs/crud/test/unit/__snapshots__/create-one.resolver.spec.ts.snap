// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateOneResolver should generate graphql schema for CreateOneCompanyResolver 1`] = `
"type User {
  id: String!
  name: String!
}

type Company {
  id: String!
  name: String!
  description: String
  founder: User!
  ceo: User
  cto: User
}

type Mutation {
  createOneCompany(data: CompanyCreateInput): Company!
}

input CompanyCreateInput {
  name: String
  founder: UserConnectNestedOneWithoutCompanyInput
}

input UserConnectNestedOneWithoutCompanyInput {
  connect: ConnectRelationInput
}

input ConnectRelationInput {
  id: String!
}"
`;

exports[`CreateOneResolver should generate graphql schema for CreateOneGroupResolver 1`] = `
"type User {
  id: String!
  name: String!
}

type Group {
  id: String!
  name: String!
  description: String!
  founders: [User!]!
  coordinator: [User!]
  finance: [User!]
}

type Mutation {
  createOneGroup(data: GroupCreateInput): Group!
}

input GroupCreateInput {
  name: String
  description: String
  founders: UserConnectNestedManyWithoutGroupInput
}

input UserConnectNestedManyWithoutGroupInput {
  connect: [ConnectRelationInput!]
  disconnect: [EntityIdInput!]
}

input ConnectRelationInput {
  id: String!
}

input EntityIdInput {
  id: String!
}"
`;

exports[`CreateOneResolver should generate graphql schema for CreateOneUserResolver 1`] = `
"type User {
  id: String!
  name: String!
}

type Mutation {
  createOneUser(data: UserCreateInput): User!
}

input UserCreateInput {
  name: String
}"
`;
